In [27]:
import pandas as pd
import numpy as np

#### Question 1
What's the version of Pandas that you installed?

In [20]:
pd.__version__

'2.1.0'

Answer: '2.1.0'

#### Getting the data

In [3]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv

--2023-09-19 16:36:01--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘housing.csv’

housing.csv         100%[===================>]   1.36M   623KB/s    in 2.2s    

2023-09-19 16:36:05 (623 KB/s) - ‘housing.csv’ saved [1423529/1423529]



#### Question 2
How many columns are in the dataset?

In [21]:
df = pd.read_csv("./housing.csv")
df.shape

(20640, 10)

Answer: 10

### Question 3
Which columns in the dataset have missing values?

In [22]:
df.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

Answer: total_bedrooms

#### Question 4
How many unique values does the ocean_proximity column have?

In [23]:
df.ocean_proximity.unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

Answer: 5

#### Question 5
What's the average value of the median_house_value for the houses located near the bay?

In [24]:
houses_near_bay = df[df['ocean_proximity'] == 'NEAR BAY']
houses_near_bay.median_house_value.mean()

259212.31179039303

Answer: 259212

#### Question 6
1. Calculate the average of total_bedrooms column in the dataset.
2. Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
3. Now, calculate the average of total_bedrooms again.
4. Has it changed?

In [25]:
# 1
bedrooms_mean = round(df.total_bedrooms.mean(), 3)
bedrooms_mean

537.871

In [26]:
# 2
df.total_bedrooms.fillna(bedrooms_mean, inplace=True)
df.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64

In [18]:
# 3
round(df.total_bedrooms.mean(), 3)

537.871

Answer: No

### Question 7
1. Select all the options located on islands.
2. Select only columns housing_median_age, total_rooms, total_bedrooms.
3. Get the underlying NumPy array. Let's call it X.
4. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
5. Compute the inverse of XTX.
6. Create an array y with values [950, 1300, 800, 1000, 1300].
7. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
8. What's the value of the last element of w?

In [28]:
# 1
islands_data = df[df['ocean_proximity'] == 'ISLAND']

In [29]:
# 2
selected_columns = ['housing_median_age', 'total_rooms', 'total_bedrooms']
selected_data = islands_data[selected_columns]

In [30]:
# 3
X = selected_data.to_numpy()

In [31]:
# 4
XTX = np.dot(X.T, X)

In [32]:
# 5
XTX_inverse = np.linalg.inv(XTX)

In [33]:
# 6
y = np.array([950, 1300, 800, 1000, 1300])

In [35]:
# 7
w = np.dot(np.dot(XTX_inverse, X.T), y)

In [36]:
# 8
last_element_of_w = w[-1]

In [37]:
last_element_of_w

5.699229455065586

Answer: 5.6992